README.md

```diff
@@ -16,10 +16,18 @@ Note: detailed go documentation is available at [godoc.org](https://godoc.org/gi
 ### Concepts
 `go-diskfs` has a few basic concepts:
 
+* Backend
 * Disk
 * Partition
 * Filesystem
 
+#### Backend
+Backend is a (relatively) thin layer which abstracts low-level read/write operations. Through a backend you can seamlessly operate different disk formats (see a bit stale qcow2 branch).
+
+Currently there is only one implementation - raw.
+
+Use `raw` backend to access block devices and raw image files.
```

Comment:
Reading this twice made me think, is "raw" the right name? Or should the backend just be `file` or `path`? That would simplify the interface to `file.Open()` and `file.Create()` (or `path.Open()` and `path.Create()`).

Reply:
> Reading this twice made me think, is "raw" the right name?

I was thinking of naming backends according to supported image formats. So, raw backend supports raw images. The next one will be, probably, qcow2.

Do you think, it is misleading?

Reply:
But qcow2 is a filesystem, which is lower-level than a disk. You get to qcow2 by doing `diskfs.Open()` (or `Create()`), which returns a handle to a `Disk`. What you do with that - reading a partition table, or a filesystem, or creating one, or partition with a filesystem on one of the partitions, is all at a next level down. `Disk` doesn't care or know about what the internal structure is.

Reply:
> But qcow2 is a filesystem, which is lower-level than a disk

Correct, it is on the same level as raw backend is!

> You get to qcow2 by doing diskfs.Open()

Something like `diskfs.OpenBackend(qcow2.New("./the_image"))`, right? Disk does not care or know about what the internal structure is, indeed!

Reply:
Yeah, but I don't think you need a separate backend for qcow2. Backend does not know or think about the contents. 

```
---> Disk (wraps) ---> Backend (handles) ---> low-level i/o
```

Dealing with qcow2 is handled at a higher level, not at a lower level. If you look at the qcow2 branch, it just changes how `Disk` deals with its writes. Adding your (good) `Backend` abstraction shouldn't change that. It just abstracts out i/o; what gets written to it is handled in `Disk` (or a level above).

To use your example, I think it would be more `d := diskfs.OpenBackend(file.New("./the_image"))`. Inside of `d` we would have ways of handling the raw vs qcow2. I don't think it goes in `Backend`. I may discover otherwise as I pick up qcow2 again, but I don't think so.

Reply:
Well, to be frank, I did not dug deep into qcow2 branch, but I saw that it has its own implementation of [ReadAt](https://github.com/aol-nnov/go-diskfs/blob/revive-qcow2/backend/qcow2/qcow2.go#L136) and other methods from backend.WritableFile, so I thought to myself, that it'd be a perfect match!

Nice separation of concerns - `backend` abstracts i/o and `Disk` (on the next level) handles partitions, filesystems, etc. without knowing anything about underlying storage. This way we do not introduce any backend implementation details to `Disk` at all.

I might be wrong, though..

Reply:
> Nice separation of concerns

Exactly. Which is why I think qcow2 is at a higher level. `Backend` is just i/o, it doesn't know if it is raw or qcow2 or something else entirely.

Reply:
I won't argue for long )) , but my final thoughs are this: look at how it is organised in linux - you can mount raw images with loop mount and qcow2 images, say, via fuse. No portions of kernel dedicated to partitions and filesystems are aware of qcow2 format. This knowledge is encapsulated in the lower level.

If you're still not convinced, I'll remove lines in question tomorrow, as I'm leaving for today.

I'm really not much aware of your plans regarding the future, so no pushing from my side - just thinking aloud!

Reply:
> won't argue for long

Why not? It keeps life interesting. Arguing with smart people is one of the great intellectual joys of life.

I do understand your point. Mostly, in my head, I am mapping it to how we work now. That is, "here is a file path (disk image or block device), get a reference `Disk` to it, and then we can use that to get the internal structure of it."

How about we rename `raw` to `file` for now, assume it is the only way. If we have to redo it to support qcow2 in the future, as we go back to it, we can. I think it will be the same backend, maybe with different properties.

Reply:
> Arguing with smart people is one of the great intellectual joys of life.

This FLUTTER-s me, so I'm GO-ing to add some more ))) (all pun intended, ha!)

>  I think it will be the same backend, maybe with different properties.

This bothers me a bit, to be frank... It's a bit against the beloved "unix way", when each part performs one simple task.

What if, by the time passes, you decide to add more formats? Say, vmdk or anything else? (Disclamer: I'm not well aware of those formats internals) Everything will be placed inside that `file` backend with different properties?

To sum up, I tend to keep things separate: raw bachend as a reference implementation and more complex implementations alongside of it.

Moreover, I had some thoughts about factoring ext4, fat and iso out of this repo and to establish an API for registering external implementations to allow users to use their own.

I understand, that implementing such things in users-space is quite a niche task and only a few users would benefit from such a change, but it would lighten the core of this project, too.

Still not convinced? Then, I'll obey you as BDFL of this project ))

Reply:
Also, if you'd like to bring some magic (as in `man 5 magic`) to users by automatically selecting the backend based on provided file image format, we could implement such a selector, too.

Reply:
> bring some magic

Do you mean something like [this magic](https://github.com/deitch/magic)? Haven't touched it in some time, but still use it in a few places. 😁 

> bit against the beloved "unix way"

I disagree, I think it is precisely the Unix way. Backend handles one thing, and one thing only: interfacing with a storage device. It is a "storage driver", not a "format driver". Following the unix way, as you suggested, I could see us splitting a "file" backend from a "block device" backend (or device driver, etc.), i.e. difference between `/home/aol-nnv/disk.img` and `/dev/sd0`. Right now, they are really close, we just have a bit of logic to distinguish between them; that would be easier if we had different backends. Another might be `memory` (for really small ones).

If we did vmdk (I looked at it at some point, really remember nothing about it), like qcow2, these would be a level above, right now in `Disk`. Also, I like that idea, although vmdk is becoming less and less interesting, as VMWare itself appears to be going through some challenges post-acquisition. On the other hand, demand for utilities to handle those may grow. Who knows?

> had some thoughts about factoring ext4, fat and iso out of this repo and to establish an API for registering external implementations to allow users to use their own

Feel free to open an iissue to discuss it; I want to see where your thoughts are going.

> Still not convinced? 

Nope, not really. I can see how in the future we might want some layer to distinguish between disk types, beyond having them part of `Disk` itself. I don't think those will be `Backend` - which will be purely i/o drivers, sort of "layer 2" (OSI model) - but either part of `Disk` or something else yet to be discovered.

Of course, we only will know for sure once we get there.

Reply:
> Do you mean something like this magic? 

Yep, exactly)

> What is "BDFL"?

https://en.m.wikipedia.org/wiki/Benevolent_dictator_for_life

I thought, it's an established term among programmers 😁

Okay, I'll try to find a few minutes tomorrow to finalize this. Turned into a bit of a journey already)) (and I'm not saying I didn't like it!)